avae/decoders/differentiable.py

```diff
@@ -0,0 +1,340 @@
+        default_axis: CartesianAxes = CartesianAxes.Z,
+        device: torch.device = torch.device("cpu"),
+        pose_dims: int,
+    ):
+        super().__init__()
+
+        self._device = device
+        # centroids should be in the range of (-1, 1)
+        self.centroids = torch.nn.Sequential(
+            torch.nn.Linear(latent_dims, n_splats * 3),
+            torch.nn.Tanh(),
+        )
+        # weights are effectively whether a splat is used or not
+        # use a soft step function to make this `binary` (but differentiable)
+        # NOTE(arl): not sure if this really makes any difference
+        self.weights = torch.nn.Sequential(
+            torch.nn.Linear(latent_dims, n_splats),
+            torch.nn.Tanh(),
+            SoftStep(k=10.0),
+        )
+        # sigma ends up being scaled by `splat_sigma_range`
+        self.sigmas = torch.nn.Sequential(
+            torch.nn.Linear(latent_dims, n_splats),
+            torch.nn.Sigmoid(),
+        )
+        # now set up the differentiable renderer
+        self.configure_renderer(
+            shape,
+            splat_sigma_range=splat_sigma_range,
+            default_axis=default_axis,
+            device=device,
+        )
+
+        self._device = device
+        self._ndim = len(shape)
+        self._output_channels = output_channels
+
+        """Decode the splats to retrieve the coordinates, weights and sigmas."""
+        if pose_dims not in (1, 4):
+            raise ValueError(
+                "Pose needs to be either a single angle rotation about the "
+                "`default_axis` or a full angle-axis representation in 3D. "
+            )
+        self.pose = not (pose_dims == 0)
+
+        # add a final convolutional decoder to generate an image if the number
+        # of output channels has been provided
+        if output_channels is not None:
+            conv = (
+                torch.nn.Conv2d
+                if self._ndim == SpatialDims.TWO
+                else torch.nn.Conv3d
+            )
+            self._decoder = torch.nn.Sequential(
+                conv(1, 32, 3, padding="same"),
+                torch.nn.ReLU(),
+                conv(32, 32, 3, padding="same"),
+                torch.nn.ReLU(),
+                conv(32, output_channels, 3, padding="same"),
+            )
+
+    def configure_renderer(
+        self,
+        shape: Tuple[int],
+        *,
+        splat_sigma_range: Tuple[float, float] = (0.02, 0.1),
+        default_axis: CartesianAxes = CartesianAxes.Z,
+        device: torch.device = torch.device("cpu"),
+    ) -> None:
+        """Reconfigure the renderer.
+
+        Notes
+        -----
+        This might be useful to do once a model is trained. For example, one
+        could change the resolution of the rendered image by changing the
+        `shape` of the output.
+        """
+        self._shape = shape
+        self._default_axis = default_axis.as_tensor()
+        self._splatter.configure_renderer(
```

Comment:
i've tried to run this model and I get an error here because `self._splatter` has not been defined anywhere:

```File "/Users/crangelsmith/Projects/affinity-vae/avae/decoders/differentiable.py", line 253, in configure_renderer
    self._splatter.configure_renderer(
  File "/Users/crangelsmith/Projects/affinity-vae/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1614, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'GaussianSplatDecoder' object has no attribute '_splatter'
```
are we missing a line around here?

Reply:
This bug has been introduced when I was dealing with mypy, in the next commit I have resolved it but pre-commit fails complaining that the about the gaussiansplatrenderer not being compatible with AbstractDecoder.

Reply:
I have put this as a temporary fix in case you need the branch until I look into the mypy error.

Reply:
This might help understand what is happening with the `mypy` error https://stackoverflow.com/questions/54346721/mypy-argument-of-method-incompatible-with-supertype

Reply:
I've fixed this issue by making the `GaussianSplatRenderer` inherit from `nn.torch.Module` instead of the abstract class.

Reply:
Thanks @crangelsmith